designs/2020-generic-ast-support/README.md

```diff
@@ -0,0 +1,57 @@
+ - nodeIdentifier: string - Specifies name of the ASTNode property that identifies node. Defaults to `type` for backwards compatibility.
+ - supportCodePathAnalysis: Boolean - Indicates AST's support for built-in CodePathAnalysis support. Defaults to `true` for backwards compatibility.
+ - supportScope: Boolean - Indicates AST's support for scope analysis. Defaults to `true` for backwards compatibility.
+ - rootNode: string - Provides the name of the top level node in the AST. Defaults to `Program` for backwards compatibility.
+
+ This would change would allow basic support of linting non-estree compliant ASTs. This will still not enable full feature parity with ESLint's abilities when linting JavaScript. In the future new properties might be added to support parsing comments to enable inline directives for controlling rules. Other enhancements might include ability to provide replacement for CodePathAnalysis and scopes creation.
+
+## Documentation
+
+This will be documented as part of the page that describes [working with custom parsers](https://eslint.org/docs/developer-guide/working-with-custom-parsers). In addition, blog post would let other users know that the can start creating parsers and rules for new languages. Although, it might be a good idea to wait until there's an example in the wild that could be pointed to as a reference.
+
+## Drawbacks
+
+The only drawback that I can think of, is that JavaScript specific rules will not work for other ASTs. Each language will most likely have to implement their own set of rules, and they will be incompatible across different parsers. While this might cause some confusion for users, I do believe this makes sense and there is no need to try to somehow generalize rules to support multiple ASTs.
+
+## Backwards Compatibility Analysis
+
+This change will be fully backwards compatible. In the future, default described above could be removed and moved into espree instead.
+
+## Alternatives
+
+None that I can think of. Other then not to support any ASTs other then estree compliant.
+
+## Open Questions
+
+* Is it OK to release incomplete support for other AST types? Inline directives being the biggest obstacle, from my perspective.
+* Should `nodeIdentifier` be passed into rules through `context` to allow creations of very generic rules that can support multiple languages/ASTs?
+* Should additional metadata be added to the `parserSupportOptions` to indicate what languages a given parser supports?
```

Comment:
This question sounds similar to the v1 processors API where processors specified what languages they supported. I prefer the v2 API where users are responsible for configuring whatever processor on whatever file, which would argue that we don't need to do this.

Reply:
Oh, sorry, wasn't clear on this question, I didn't mean it as parser would drive inclusion of files (or not), I meant more from a metadata perspective. For example, HTML formatter might want to break down percentage of errors from a lint run based on the language, or something like that.

Reply:
Ah! In that case I could see it as being speculatively useful, and it'd be easier to do from the very beginning rather than adding retroactively. I'm curious to see what other use cases people can come up with for this.